cli/tools/vendor/mappings.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::path::Path;
+use std::path::PathBuf;
+
+use deno_ast::MediaType;
+use deno_ast::ModuleSpecifier;
+use deno_core::error::AnyError;
+use deno_graph::Module;
+use deno_graph::ModuleGraph;
+use deno_graph::Position;
+use deno_graph::Resolved;
+
+use crate::fs_util::path_with_stem_suffix;
+
+use super::specifiers::dir_name_for_root;
+use super::specifiers::get_unique_path;
+use super::specifiers::make_url_relative;
+use super::specifiers::partition_by_root_specifiers;
+use super::specifiers::sanitize_filepath;
+
+pub struct ProxiedModule {
+  pub output_path: PathBuf,
+  pub declaration_specifier: ModuleSpecifier,
+}
+
+/// Constructs and holds the remote specifier to local path mappings.
+pub struct Mappings {
+  output_dir: ModuleSpecifier,
+  mappings: HashMap<ModuleSpecifier, PathBuf>,
+  base_specifiers: Vec<ModuleSpecifier>,
+  proxies: HashMap<ModuleSpecifier, ProxiedModule>,
+}
+
+impl Mappings {
+  pub fn from_remote_modules(
+    graph: &ModuleGraph,
+    remote_modules: &[&Module],
+    output_dir: &Path,
+  ) -> Result<Self, AnyError> {
+    let partitioned_specifiers =
+      partition_by_root_specifiers(remote_modules.iter().map(|m| &m.specifier));
+    let mut mapped_paths = HashSet::new();
+    let mut mappings = HashMap::new();
+    let mut proxies = HashMap::new();
+    let mut base_specifiers = Vec::new();
+
+    for (root, specifiers) in partitioned_specifiers.into_iter() {
+      let base_dir = get_unique_path(
+        output_dir.join(dir_name_for_root(&root)),
+        &mut mapped_paths,
+      );
+      for specifier in specifiers {
+        let media_type = graph.get(&specifier).unwrap().media_type;
+        let sub_path = sanitize_filepath(&make_url_relative(&root, &{
+          let mut specifier = specifier.clone();
+          specifier.set_query(None);
+          specifier
+        })?);
+        let new_path = path_with_extension(
+          &base_dir.join(if cfg!(windows) {
+            sub_path.replace('/', "\\")
+          } else {
+            sub_path
+          }),
+          &media_type.as_ts_extension()[1..],
+        );
+        mappings
+          .insert(specifier, get_unique_path(new_path, &mut mapped_paths));
+      }
+      base_specifiers.push(root.clone());
+      mappings.insert(root, base_dir);
+    }
+
+    // resolve all the "proxy" paths to use for when an x-typescript-types header is specified
+    for module in remote_modules {
+      if let Some((
+        _,
+        Resolved::Ok {
+          specifier, range, ..
+        },
+      )) = &module.maybe_types_dependency
+      {
+        // hack to tell if it's a x-typescript-types header
```

Comment:
I don't get what this hack does :(

Reply:
See https://github.com/denoland/deno_graph/issues/133

The hack is to check if the range start and end in the file is zero.